Python support on Literals

dramaticlly opened the pull request:
First step of closing https://github.com/apache/iceberg/issues/4250

This is my first iceberg python contribution, so there might be some naive mistake and please let me know so I can jump to fix, most of the code is refactoring from https://github.com/apache/iceberg/blob/master/python_legacy/iceberg/api/expressions/literals.py to our latest python module

Change highlights
- Refactor all static method to module level methods to be more pythonic
- Use [singledispath](https://docs.python.org/3/library/functools.html#functools.singledispatch) to dynamic overload the Literal factory from python primitive type and conversion between iceberg types
- Adopt the new TimestampType from types, 
   - replace `TimestampType.with_timezone` with `TimestampTzType` 
   - replace `TimestampType.without_timezone` with `TimestampType` 
- Combined 3 literal unit tests together with parametrized pytest style python unit tests

dramaticlly commented:
@jun-he , do you mind give it a first pass, I want to make sure if this looks ok to you as this is my first python PR

dramaticlly commented on python/src/iceberg/literals.py:
```diff
@@ -0,0 +1,580 @@
+        return BELOW_MIN
+
+
+def from_(value):  # noqa: C901
+    if value is None:
+        raise RuntimeError("Cannot create an expression literal from None")
+    if isinstance(value, bool):
+        return BooleanLiteral(value)
+    elif isinstance(value, int):
+        if Literal.JAVA_MIN_INT < value < Literal.JAVA_MAX_INT:
+            return IntegerLiteral(value)
+        return LongLiteral(value)
+    elif isinstance(value, float):
+        if Literal.JAVA_MIN_FLOAT < value < Literal.JAVA_MAX_FLOAT:
+            return FloatLiteral(value)
+        return DoubleLiteral(value)
+    elif isinstance(value, str):
+        return StringLiteral(value)
+    elif isinstance(value, uuid.UUID):
+        return UUIDLiteral(value)
+    elif isinstance(value, bytearray):
+        return BinaryLiteral(value)
+    elif isinstance(value, bytes):
+        return FixedLiteral(value)
+    elif isinstance(value, Decimal):
+        return DecimalLiteral(value)
+    else:
+        raise NotImplementedError("Unimplemented Type Literal for value: %s" % value)
```
as moved to python module methods, this looks very similar to LINE81 to 101, so I am thinking about merge them so we can have only 1 factory.

samredai commented on python/src/iceberg/literals.py:
```diff
@@ -0,0 +1,595 @@
+        return not self.__eq__(other)
+
+    def __repr__(self):
+        return "BaseLiteral(%s)" % str(self.value)
```
Is it true that we don't expect BaseLiteral to ever be instantiated directly? If so maybe we should remove this `__repr__` and instead add `__repr__` methods directly to the subclasses like `BooleanLiteral` and `IntegerLiteral`. Or you can take the same approach as [types.py](https://github.com/apache/iceberg/blob/master/python/src/iceberg/types.py#L49-#L58) where the repr and str methods are defined as part of the super init.

samredai commented on python/src/iceberg/literals.py:
```diff
@@ -0,0 +1,595 @@
+    elif isinstance(value, bytearray):
+        return BinaryLiteral(value)
+    elif isinstance(value, Decimal):
+        return DecimalLiteral(value)
```
I'd suggest `@singledispatch` here to enable overloading. If you need an example I just opened PR #4263 which uses that.

dramaticlly replied:
yeah I just saw that, thank you Sam for your great example!

samredai commented on python/src/iceberg/literals.py:
```diff
@@ -0,0 +1,595 @@
+        return self.value >= other.value
+
+
+class FixedLiteralProxy(object):
```
`(object)` can be removed here and in a few other places.

dramaticlly replied:
will do

samredai commented on python/src/iceberg/literals.py:
```diff
@@ -0,0 +1,595 @@
+)
+
+
+class Literals(object):
```
I'm going to spend more time looking at this but is there a reason why we need both `Literal` and `BaseLiteral`? Could what's here be collapsed into `BaseLiteral`?

dramaticlly replied:
after some looking, I believe this is the hierarchy graph

```
Literal
|-- AboveMax
|-- BelowMin
|-- BaseLiteral
    |-- StringLiteral
    |-- FixedLiteral    
    |-- BinaryLiteral        
    |-- ComparableLiteral
        |-- BooleanLiteral
        |-- IntegerLiteral
        |-- LongLiteral
        |-- FloatLiteral
        |-- DoubleLiteral
        |-- DateLiteral
        |-- TimeLiteral
        |-- TimestampLiteral
        |-- DecimalLiteral
        |-- UUIDLiteral
```

samredai commented on python/src/iceberg/literals.py:
```diff
@@ -0,0 +1,595 @@
+
+    def to_byte_buffer(self):
+        if self.byte_buffer is None:
+            # TODO need byte buffer conversion
```
I just opened PR #4263 for this!

dramaticlly replied:
what a great timing!

samredai commented on python/src/iceberg/literals.py:
```diff
@@ -0,0 +1,595 @@
+        return FixedLiteral(self.bytes)
+
+
+class ConstantExpressionProxy(object):
```
@TGooch44 (or anyone else)--sorry if this is obvious but what's the purpose of these "Proxy" classes?

dramaticlly replied:
Since multiple reviewer suggested, I end up removing such code as it's not being used. Once we get needs and clarify, we can always add them later if needed. So resolve this

samredai commented on python/setup.py:
```diff
@@ -19,7 +19,7 @@
 setup(
     name="py-iceberg",
-    install_requires=[],
+    install_requires=["pytz", "python-dateutil", "fastavro>=1.3.2<1.4.0", "python-dateutil"],
```
I believe these should be added to the `pyproject.toml` file instead of here.

samredai reviewed (commented):
@dramaticlly I left some comments. I think there are some things here that we can refactor/clean up from the legacy library.

One thing I noticed is that I can't quickly determine where the concrete classes are as opposed to base classes. I'm wondering if the library would be more navigable if we maintain a pattern where we section off interfaces into separate files away from the concrete classes with some exceptions of course.

Since literals are only used in expressions and there are a lot more expression related components we can expect soon, maybe we can create an `expressions` directory and move `literals.py` there. In a separate PR we can move+rename `expressions.py` to `expressions/operations.py` since that's all it contains so far.

We can then move all of the base classes to `base.py` and add some clarity by making them true ABC classes, leaving the other files to only contain concrete classes.

```
expressions
├── __init__.py
├── base.py
├── literals.py
└── operations.py
```

dramaticlly commented on python/src/iceberg/literals.py:
```diff
@@ -0,0 +1,595 @@
+    JAVA_MAX_INT = 2147483647
+    JAVA_MIN_INT = -2147483648
+    JAVA_MAX_FLOAT = 3.4028235e38
+    JAVA_MIN_FLOAT = -3.4028235e38
```
Sounds good Sam, I can move this to module level constant

dramaticlly commented on python/src/iceberg/literals.py:
```diff
@@ -0,0 +1,595 @@
+            return self
+        elif isinstance(type_var, UUIDType):
+            return UUIDLiteral(uuid.UUID(self.value))
+        elif isinstance(type_var, DecimalType):
```
yep, I will look into the singledispatchmethod and type based registration

dramaticlly replied:
Looks like the [@singledispatchmethod](https://docs.python.org/3/library/functools.html#functools.singledispatchmethod) helps but only available in python 3.8. So this actually make our tox CI on 3.7 fails. As much as I want to keep this neat way of dynamic overloading, do we know if we want to support minimally for 3.7?

If so, I guess we cannot use it within the class method

dramaticlly commented on python/src/iceberg/literals.py:
```diff
@@ -0,0 +1,595 @@
+
+
+ABOVE_MAX = AboveMax()
+BELOW_MIN = BelowMin()
```
Yeah that make sense to me, thank you for your example code snippet

dramaticlly commented on python/tests/test_string_literal_conversions.py:
```diff
@@ -0,0 +1,104 @@
+from decimal import Decimal
+
+import dateutil.parser
+from fastavro.write import LOGICAL_WRITERS as avro_conversion
```
sounds good to me, will do

dramaticlly commented on python/tests/test_misc_literal_conversions.py:
```diff
@@ -0,0 +1,308 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
yeah that was my original intention as well, then I realized there's so many test cases for this source code so I ends up bring each and everyone of them. If you prefer to merge them into a single `test_literals.py` I can do that!

samredai commented:
cc: @CircArgs

kbendick commented on python/tests/test_misc_literal_conversions.py:
```diff
@@ -0,0 +1,308 @@
+
+
+def test_invalid_fixed_conversions():
+    if sys.version_info[0] >= 3:
```
Are we supporting Python 2.7?

I know that in some respects, Python 2.7 won't die. But if we don't have to support it, that would be great.

Otherwise, does the `setup.py` file or anything else provide us a means of indicating required Python version?

dramaticlly replied:
Thank you @kbendick , I think that's a really good callout. 

I did look around existing code in the new module and find nothing to limit the minimal python version we support. I think if we plan to only support it for 3 onwards (with some helpful python feature we can use as part of development), we shall do it. I think that's the topic warrant the discussion on next Python meet.

Once we decides, I can go over the code base and fix the check like this, what do you think?

samredai replied:
Good catch @kbendick, I think there was an agreement early on to support python 3.7+ and it's set in the [setup.cfg](https://github.com/apache/iceberg/blob/master/python/setup.cfg#L44) file. The current state of python packaging feels like it's in flux (maybe it always has been?) and it's hard to tell where the right place to set this stuff is. I **think** the current standard is to use the pyproject.toml+setup.cfg combo...

@dramaticlly for these if statements you can definitely replace them with the assumption that python 3 is being used.

dramaticlly replied:
Sounds good, thank you Sam for the clarification. With that being understood, I will remove those python version checks :)

szehon-ho commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,737 @@
+@of.register
+def _of(value: int):
+    """
+    Upgrade to long is python int is outside the JAVA_MIN_INT and JAVA_MAX_INT
```
Reading the sentence a few times, is the first "is" a typo and should be "as"?

dramaticlly replied:
yes that's typo, meant to say `Upgrade to long if python int is outside the JAVA_MIN_INT and JAVA_MAX_INT`

szehon-ho commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,737 @@
+@singledispatch
+def of(value):
+    """
+    A generic Literal factory to construct a iceberg Literal based on python primitive data type
```
super nit: "an Iceberg Literal"?

samredai commented on python/.coveragerc:
```diff
@@ -20,5 +20,5 @@ skip_empty = true
 omit =
-    # omit this single file
-    src/iceberg/literals.py
+    src/iceberg/expression/literals.py
```
nit: should we make the directory name plural, as in `expressions`? I'm thinking of how there are different types of expressions that will be defined here like And, Or, Not etc...

dramaticlly replied:
ops, yes you are right @samredai , the only problem with that is that [expressions.py](https://github.com/apache/iceberg/blob/master/python/src/iceberg/expressions.py#L21) already exists so it will conflict when I try to import is as module. As you suggested, I plan to rename the file in a separate PR to get the directory naming correctly as what you suggested, what do you think?

samredai replied:
Ah right, I think we should rename that file to "operations.py" since it currently only contains stuff related to operations.

jun-he replied:
Given that no other places use literals, I am wondering if we can remove those unfinished code (e.g. todos) from this PR and move them into another PR? So we can still achieve the same level of code coverage for merged code.

dramaticlly replied:
@jun-he , I understand that from logical perspective, it make sense to have consistent standard where all merged codes are conforming to the same requirement on code coverage. But in this specific Literal cases, I found many places of coverage gap comes from base base such as 

- BaseLiterals
<img width="767" alt="image" src="https://user-images.githubusercontent.com/5961173/157098041-e3ded4f6-466f-4b04-b159-91b982b36819.png">

- ComparableLiterals
<img width="644" alt="image" src="https://user-images.githubusercontent.com/5961173/157098129-5d7c1d13-3349-477a-8ac6-0f87119c10f3.png">

Which if we remove the code for those, then it does not make sense for rest of the code to check in. So there's no very clean way of separating them out. What we can do here is, either ask for more code coverage for internal method of base class, to merge it within same PR, or allow for some temporary exclusion and assign the specific issue and come back in a separate PR.

samredai replied:
> yep I agree, do you think it would be better to do the renaming in separate PR?

Yes that sounds good to me. It's an easy follow-up PR. 👍

samredai replied:
re: coverage--If all of this logic in the base classes are not used anywhere, we shouldn't need them at all right? I think defining these as ABC classes will allow us to provide more clarity by clearly decorating what's an `@abstractmethod` and leaving the function body empty (just a doc string). That's a clear signal to coverage as well that there's nothing to cover. Any concrete methods in an ABC class would be there because they're used somewhere which means they can be covered in tests. Similar to what's in [io/base.py](https://github.com/apache/iceberg/blob/master/python/src/iceberg/io/base.py).

szehon-ho commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,737 @@
+@of.register
+def _of(value: float):
+    """
+    Upgrade to double is python float is outside the JAVA_MIN_FLOAT and JAVA_MAX_FLOAT
```
Same as above

szehon-ho commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,737 @@
+
+
+class BaseLiteral(Literal):
+    """Base literal which have a value and can be converted between types"""
```
super nit: "literal which has"

samredai commented on python/pyproject.toml:
```diff
@@ -18,6 +18,10 @@
 requires = [
     "setuptools>=42",
-    "wheel"
+    "wheel",
```
You should also add `singledispatch` here which includes the backport of `@singledispatchmethod` for python 3.7 (adding it to tox.ini also should fix the test failure in your latest commit)

dramaticlly replied:
Thank you @samredai  for your suggestion for py37 `singledispatch` backport, I managed to get it working. But mypy is still unable to redefining the same method and I double checked it's on the [latest 0.931](https://github.com/apache/iceberg/runs/5430406474?check_suite_focus=true#step:5:369) as you suggested. I am scratching my head trying to see what's the correct way to make it understand this is not an error,

dramaticlly replied:
I think I can temporarily silence the mypy linter error with `# type: ignore[no-redef]` for each method I am overriding. In the meantime I can read more with how mypy works with those

samredai replied:
I looked a bit closer at what mypy is doing and it's getting confused because you've given actual names to the registered functions. The convention is to only name the generic function and use `_` for the additional type based functions that you're registering (like the example in the [docs](https://docs.python.org/3/library/functools.html#functools.singledispatchmethod)). I confirmed that will pass all mypy inspection.

samredai replied:
To my comment about adding `singledispatch` and the other dependencies here, I did a little more research and `install_requires` in [setup.cfg](https://github.com/apache/iceberg/blob/master/python/setup.cfg#L40) is actually the right place for this. My bad!
```
[options]
...
install_requires =
    pytz
    python-dateutil
    fastavro>=1.3.2<1.4.0
    python-dateutil
    singledispatch
```

samredai commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,741 @@
+try:
+    from functools import singledispatchmethod
+except ImportError:
+    from singledispatch import singledispatchmethod  # type: ignore
```
I think it'd be helpful if we're more deliberate here that we're falling back to the backport for versions below 3.8, that way in the future when we drop python 3.7 support it will be obvious that this line can be removed. How about:
```py
import sys

if sys.version_info >= (3, 8):
    from functools import singledispatchmethod
else:
    from singledispatch import singledispatchmethod
```

jun-he commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,741 @@
+    UUIDType,
+)
+
+JAVA_MAX_INT = 2147483647
```
Wondering if those limits are defined in iceberg spec? If so, we may rename those constants to `ICEBERG_...` instead of `JAVA_...` to make it clear.

dramaticlly replied:
I was looking at https://iceberg.apache.org/spec/ and did NOT find any explicit mentioning of it, I guess this comes from `Integer.MAX_VALUE` so maybe we can keep them as is?

jun-he commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,741 @@
+class BaseLiteral(Literal):
+    """Base literal which has a value and can be converted between types"""
+
+    def __init__(self, repr_string: str, value):
```
type annotation for `value`?

dramaticlly replied:
for `value` here it would be typed of `Any` as well, let me know if you want me to add it

jun-he commented on python/tests/expression/conftest.py:
```diff
@@ -0,0 +1,32 @@
+
+from iceberg.types import DecimalType
+
+
```
is the conftest to set up base config for expression tests?

dramaticlly replied:
yes Jun, that's my intention. This conftest is used for expression unit tests

jun-he commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,742 @@
+
+
+class Literal:
+    def to(self, type_var):
```
Is it feasible to put type annotation for type_var here?

dramaticlly replied:
I guess if we need , the `type_var` would be of `Any` type here as it's the base method and type_var can be unrestricted and concrete Literals below will have better defined method for allowed type conversion.

samredai replied:
Would `PrimitiveType` work?

dramaticlly replied:
added the PrimitiveType, I think this is a good hint :)

jun-he commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,742 @@
+
+
+@of.register(bool)
+def _of(value):
```
Is it feasible to put type annotation for `value`?

dramaticlly replied:
there's type annotation on the decorator above, do you think if it's sufficient? or I can add for `value` if you prefer

samredai replied:
`@singledispatch` is actually clever enough to look at the typehint, so you could actually do this:
```py
@of
def _(value: bool):
    ...
```
The one downside is that if you want to dispatch one function for multiple types, you can't stack them the way you can with the decorator that calls register and unfortunately it doesn't recognize the `typing.Union` type hint where you specify multiple types. That's not a concern here though since it's a 1:1 mapping between types and logic.

dramaticlly replied:
yep, I ended up with moving type annotation to parameter instead of decorator since they are 1:1 mapping as of now

samredai commented on python/tests/expression/test_literals.py:
```diff
@@ -0,0 +1,479 @@
+# MISC
+
+
+def test_identity_conversions():
```
parametrize from pytest would be better here since it will create a test for every pair. Something like:
```py
@pytest.mark.parametrize(
    "literal, primitive_type",
    [
        (iceberg.expression.literals.of(True), BooleanType()),
        (iceberg.expression.literals.of(34), IntegerType()),
        (iceberg.expression.literals.of(3400000000), LongType()),
        (iceberg.expression.literals.of(34.11), FloatType()),
        (iceberg.expression.literals.of(34.11), DoubleType()),
        (iceberg.expression.literals.of(Decimal(34.55).quantize(Decimal("0.01"))), DecimalType(9, 2)),
        (iceberg.expression.literals.of("2017-08-18"), DateType()),
        (iceberg.expression.literals.of("14:21:01.919"), TimeType()),
        (iceberg.expression.literals.of("2017-08-18T14:21:01.919"), TimestampType()),
        (iceberg.expression.literals.of("abc"), StringType()),
        (iceberg.expression.literals.of(uuid.uuid4()), UUIDType()),
        (iceberg.expression.literals.of(bytes([0x01, 0x02, 0x03])), FixedType(3)),
        (iceberg.expression.literals.of(bytearray([0x03, 0x04, 0x05, 0x06])), BinaryType()),
    ],
)
def test_identity_conversions(literal, primitive_type):
    expected = literal.to(primitive_type)
    assert expected is expected.to(primitive_type)
```

dramaticlly replied:
Thank you Sam, I was also looking for any other test cases can be optimized, there's potentially some invalid type conversion but looks like the repeating pattern and we already get a helper method `assert_invalid_conversions` created for those, so I just ends up parameterize for this method you suggested.

samredai commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,742 @@
+
+
+@singledispatch
+def of(value):
```
Since this will be at the module level, is there a better name for this besides `of`? What do you think of `literal`? It feels right to do:
```py
from iceberg.expressions.literals import literal

l = literal(1.23)
print(l)  # FloatLiteral(value=1.23)
```

dramaticlly replied:
sounds good, since capital `Literal` is used, I think we can use what you suggested instead

samredai commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,742 @@
+        return None
+
+    @to.register(BooleanType)
+    def boolean_literal_to(self, type_var):
```
Just for consistency I would make name this `_` as well.

dramaticlly replied:
yeah good catch, corrected it!

dramaticlly commented on python/.python-version:
```diff
@@ -1,3 +1,3 @@
 3.7.12
 3.8.12
-3.9.10
+3.9.9
```
Actually I probably should not commit this change, but when I do the listing of python 3.9 in pyenv, somehow I cannot find the 3.9.10

```
(venv) ➜  python git:(master) ✗ pyenv install --list | grep 3.9
  3.9.0
  3.9-dev
  3.9.1
  3.9.2
  3.9.4
  3.9.5
  3.9.6
  3.9.7
  3.9.8
  3.9.9
  miniconda-3.9.1
  miniconda3-3.9.1
  miniconda3-3.9-4.9.2
  miniconda3-3.9-4.10.3
```

samredai replied:
Does `tox --recreate` help by any chance?

dramaticlly replied:
Nope, my pyenv is installed outside tox so it might be related, but I can exclude this file from PR and keep it locally to test for python 3.9

samredai commented on python/setup.cfg:
```diff
@@ -42,6 +42,12 @@ package_dir =
+    pytz
+    python-dateutil
+    fastavro>=1.3.2<1.4.0
+    python-dateutil
```
Looks like `python-dateutil` was added here twice accidentally.

dramaticlly replied:
I will fix them later together with unit testsing

CircArgs commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,745 @@
+    UUIDType,
+)
+
+JAVA_MAX_INT = 2147483647
+JAVA_MIN_INT = -2147483648
+JAVA_MAX_FLOAT = 3.4028235e38
+JAVA_MIN_FLOAT = -3.4028235e38
```
in `types.py` there exist `IntegerType.max` `IntegerType.min`, similarly for `LongType`. Perhaps we should use those and add the bounds for the floating types as well?

dramaticlly replied:
Yes, I think it make more sense to stay in types instead of literal, so I added float min and max to `types.py` and removed those `JAVA_` module level attributes

CircArgs commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,745 @@
+
+
+class AboveMax(Literal, Singleton):
+    _instance = None
```
Do we need this attribute on things inheriting from `Singleton`? It provides it already as it uses it in `__new__`

CircArgs commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,745 @@
+    _instance = None
+
+    def __init__(self):
+        self._initialized = True
```
It's not clear to me why we are setting this flag here. I don't think this is the same situation as this was used in `types.py`. There's no check for whether this flag is set here to avoid calling the super init so if this flag really is needed I think we need the check too. If we don't need the flag, I think this init overall can be gotten rid of, which I think makes sense right now

CircArgs commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,745 @@
+
+    def __init__(self):
+        self._initialized = True
+        super(AboveMax, self).__init__()
```
in general we can use the shorthand `super().__init__()` since the longhand here is the default if I'm not mistaken

dramaticlly replied:
yep, we actually don't need the `__init__` for both AboveMax and BelowMin I think

rdblue commented on python/pyproject.toml:
```diff
@@ -18,6 +18,6 @@
 ]
-build-backend = "setuptools.build_meta"
\ No newline at end of file
+build-backend = "setuptools.build_meta"
```
It doesn't look like this file substantially changed. Can we remove unnecessary changes here?

dramaticlly commented on python/setup.cfg:
```diff
@@ -42,6 +42,11 @@ package_dir =
+install_requires =
+    pytz
+    python-dateutil
+    fastavro>=1.3.2<1.4.0
```
Actually I think this is only used in unit tests so far to validate the string to date/time/timestamp conversion where its value is setup correctly. from python_legacy code base I think it has been used for manifest reader/writer so I ported it over. I can look into what's the best place to put this dependency

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,713 @@
+
+
+@singledispatch
+def literal(value: Any):
```
Should this have a return type annotation for `Literal`? And is it necessary to add those to the registered functions as well?

dramaticlly replied:
I added the return type hint for `BaseLiteral` instead,  as `value` as attribute is defined in `BaseLiteral` so I feel it make more sense to return base instead (which have value and can be convert to other Literal based on type).

The current `Literal` class is parent structure to both BaseLiteral and AboveMax/BelowMin

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,713 @@
+    def __str__(self):
+        return str(self.value)
+
+    def to_byte_buffer(self):
```
I would probably remove all of these methods until they are implemented. We can always add them later, right?

dramaticlly replied:
removed all `to_byte_buffer` and proxy class as we can add them later if needed

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,713 @@
+        return not self.__eq__(other)
+
+    def __lt__(self, other):
+        if self.value is None:
```
Literal values must never be `None`, so that we never have to worry about cases where we are comparing `None`. Can you add a `TypeError` or `ValueError` in the constructor? It is important to have this invariant for the rest of the expressions codebase.

dramaticlly replied:
gotcha, I think this is a good callout. 

In python I think it's a bit hard to enforce such invariant, there's a few way to set the attribute of class in non-standard way. So I did following thing to make it align with expectation
- add `value` with [property annotation](https://docs.python.org/3/library/functions.html#property) so setting it directly like `StringLiteral.value = None` will fail with `AttributeError: can't set attribute`
- add None check with `TypeError` in BaseLiteral constructor `__init__` call so that it will reject None on initialization
  - introduce the inner structure to `_value` as abstraction layer to store the value after initialization, in python this convey the message of private field but does not enforce it. One can still set to to None via `lit.__dict__['_value'] = None` but they are no longer covered by our code
  - `value` property getter is directly returning `_value`, so does `__str__`

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,713 @@
+
+        return self.value < other.value
+
+    def __gt__(self, other):
```
Is there an equivalent to `compareTo` that returns an `int` comparison result so we can mix in the implementations of these methods? Sometimes this is the spaceship operator, `<=>`, but I'm not sure what the equivalent is in Python.

dramaticlly replied:
Looks like python3 has [moved away from `__cmp__`](https://docs.python.org/3.0/whatsnew/3.0.html#ordering-comparisons) and instead rely on those 6 rich comparison methods
- `__eq__`
- `__ne__`
- `__lt__`
- `__gt__`
- `__le__`
- `__ge__`

Also with some basic search, I did not find anything works like spaceship operator in python3 yet. But if someone else did, I would appreciate the pointers!

dramaticlly commented:
Thank you @rdblue , I sent a new commit based on your review feedback, please let me know what do you think :)

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,654 @@
+    |-- StringLiteral
+    |-- FixedLiteral
+    |-- BinaryLiteral
+    |-- ComparableLiteral
```
Why isn't StringLiteral a ComparableLiteral? Don't the comparison functions work for it?

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,654 @@
+        else:
+            return DecimalLiteral(
+                Decimal(self.value).quantize(
+                    Decimal("." + "".join(["0" for _ in range(1, type_var.scale)]) + "1"), rounding=ROUND_HALF_UP
```
String manipulation for decimal conversion looks suspicious to me. Is there no API to do this?

dramaticlly replied:
yeah I was looking into [Decimal](https://docs.python.org/3/library/decimal.html#) and looks like the quantize is the API supposed to round the Decimal value to its first operant. I can kind of see that construct the Decimal from String is for better precision compare to float counterpart. 

I am not very familiar with the Decimal and did not find a better API to do this, I guess we can keep this as of now and come back if someone find a better way to do this ? What do you think

CircArgs replied:
perhaps instead of using this string manipulation you can use the init by tuple 
```python
Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP)`
```

dramaticlly replied:
Thank you @CircArgs , I think using tuple to initialize the Decimal looks much better than the string manipulation. Updated per your suggestion

rdblue replied:
Looks like this one didn't get updated?

rdblue replied:
I think this should be:

```python
sign, digits, _ = Decimal(self.value).to_tuple()
return Decimal((sign, digits, -type_var.scale))
```

Fokko commented on python/tox.ini:
```diff
@@ -25,6 +25,10 @@ deps =
     mock
     pytest
     pyarrow
+    pytz
```
Should we define those here as well? I would argue that we already set these in `setup.cfg`, and if we start pinning them to certain versions, it is better to set this in one place.

Fokko commented on python/tox.ini:
```diff
@@ -65,6 +69,9 @@ commands =
 [testenv:type-check]
 deps =
     mypy
+    types-pytz
```
You could also install these types implicitly: https://mypy.readthedocs.io/en/stable/command_line.html#miscellaneous
This has to be combined with the non-interactive flag: https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-non-interactive

kbendick commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,654 @@
+        raise RuntimeError("Cannot change the type of AboveMax")
+
+    def __str__(self):
+        return "aboveMax"
```
Nit: Is there a reason that this string uses `aboveMax` vs `AboveMax` like the type and the runtime error for `value` etc has?

Fokko commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,618 @@
+        raise RuntimeError("Cannot change the type of AboveMax")
+
+    def __str__(self):
+        return "AboveMax"
```
https://github.com/apache/iceberg/pull/4262#discussion_r829375822 We could also generalize this to the Literal:
```
➜  python git:(master) python3
Python 3.9.10 (main, Jan 15 2022, 11:48:04) 
[Clang 13.0.0 (clang-1300.0.29.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from iceberg.types import Singleton
>>> 
>>> class Literal:
...     def to(self, type_var):
...         raise NotImplementedError()
...     def __str__(self):
...         return type(self).__name__
... 
>>> class AboveMax(Literal, Singleton):
...     def value(self):
...         raise RuntimeError("AboveMax has no value")
...     def to(self, type_var):
...         raise RuntimeError("Cannot change the type of AboveMax")
... 
>>> str(AboveMax())
'AboveMax'
```

dramaticlly replied:
Thank you @Fokko , updated to save define `__str__` for both AboveMax and BelowMin

dramaticlly commented:
I did rebase from latest upstream and squash my commits together into a single one, have a better way to initialize the Decimal per @CircArgs  suggestion, define the `__str__` in type level `Literal` and removed the usage of `conftest.py` in pytests.

@rdblue @jun-he @samredai if you guys can take another look, that would be awesome!

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,601 @@
+    def __init__(self, repr_string: str, value):
+        self._repr_string = repr_string
+        if value is None:
+            raise TypeError("Cannot set value of BaseLiteral to None")
```
Error messages should generally not reference variable or class names. The person reading the error message is rarely familiar with this code, so using internal names is just jargon. Here, I recommend `Invalid literal value: None`.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,601 @@
+        return type(self).__name__
+
+
+class BaseLiteral(Literal):
```
I'm not sure that this class hierarchy makes sense. This class states that it has a value that can be converted between types, but the `to` method is defined in `Literal` and throw an exception. In addition, `AboveMax` and `BelowMin` are never exposed to users, so I don't see much value in a class hierarchy that excludes them from the literals that have values.

I would combine `Literal` into `BaseLiteral`.

Fokko replied:
I had the same feeling. Next to that, we could also make the BaseLiteral derive from the [abstract base class](https://docs.python.org/3/library/abc.html).

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,601 @@
+"""
+
+
+class Literal:
```
I think that this should be parameterized so you know what kind of Python object is returned by `literal.value`.

I think that should work with `@singledispatch` because it can use `def literal(value: T) -> BaseLiteral[T]:`

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,601 @@
+
+
+@singledispatch
+def literal(value: Any) -> BaseLiteral:
```
I think this should also be parameterized so you know what type the base literal contains.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,601 @@
+        super(IntegerLiteral, self).__init__(f"IntegerLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
```
Is it possible to get this to produce the correct output type based on the registered method that gets called?

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,601 @@
+        return None
+
+    @to.register(IntegerType)
+    def _(self, type_var):
```
This should return `IntegerLiteral`, right?

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,601 @@
+
+    @to.register(FloatType)
+    def _(self, type_var):
+        return FloatLiteral(self.value)
```
Should this cast to `float`?

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,601 @@
+
+
+class DoubleLiteral(ComparableLiteral):
+    def __init__(self, value):
```
I think the constructors should have type annotations so that we correctly convert values when converting literals.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,601 @@
+
+@literal.register
+def _(value: bytes):
+    return FixedLiteral(value)
```
I think that both `bytes` and `bytearray` should produce a `BinaryLiteral` that contains `bytes`. Iceberg should not hold on to `bytearray` values because they are mutable, so this should be immediately converted in the factory method.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,601 @@
+|-- BelowMin
+|-- BaseLiteral
+    |-- FixedLiteral
+    |-- BinaryLiteral
```
This PR uses `bytes` for binary, which appears to support comparison operators. That means I'd probably merge the `ComparableLiteral` methods into `Literal` as well. Everything supports normal comparison so there's no need for implementation in the type hierarchy.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,601 @@
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP))
+
+
+class FloatLiteral(ComparableLiteral):
```
I think that floats need special implementations of comparison methods like `__eq__`.

The Python VM uses 64-bit double to represent both floating point types. That means that float literals can be more exact than float values that are coming from data files. For example:
| Cell | Python code | Result | Note |
|:-- |:-- |:-- |:-- |
| 1 | `12.34` | `12.34` | Original value
| 2 | `struct.pack('<f', _1)` | `b'\xa4pEA'` | 4-byte binary |
| 3 | `struct.unpack('<f', _2)[0]` | `12.34000015258789` | Lost precision! |
| 4 | `_1 == _3` | `False` | No longer equal |
| 5 | `struct.pack('<d', _1)` | `b'\xaeG\xe1z\x14\xae(@'` | 8-byte binary |
| 6 | `struct.unpack('<d', _5)[0]` | `12.34` | Trailing 0s omitted by repr |
| 7 | `_1 == _6` | `True` | Equal when comparing extra bits |

We need to watch out for the problem shown in [3] for 32-bit float data coming from data files.

One option to fix this is to round-trip the literal value through `struct.pack` and `struct.unpack`. I think that makes the most sense, but it also raises the question: what should we pass into `DoubleLiteral` when converting? My initial take is to either not allow converting a `FloatLiteral` to `DoubleLiteral` or to pass the original python float (64-bit) that was used to create the `FloatLiteral`.

I propose we add `_value32` to this class and use that in overridden comparison methods. That way comparison acts correctly.

```python
class FloatLiteral(Literal):
    def __init__(self, value):
        super(FloatLiteral, self).__init__(f"FloatLiteral({value})", value)
        this._value32 = struct.unpack('<f', struct.pack('<f', value))[0]

    def __eq__(self, value):
        this._value32 == value

    ...
```

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,601 @@
+        return self.value < other.value
+
+    def __gt__(self, other):
+        if other is None:
```
`other` must never be `None` or else it is an error.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,601 @@
+    def _(self, type_var):
+        if type_var.length == len(self.value):
+            return FixedLiteral(self.value)
+        return None
```
This should throw an exception if conversion fails.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,601 @@
+    def _(self, type_var):
+        import dateutil.parser
+
+        return DateLiteral((dateutil.parser.parse(self.value) - EPOCH).days)
```
These conversions need to be very strict. A date must be formatted `YYYY-mm-DD`, so I think this should explicitly use a format and fail if the string deviates from the format in anyway.

Same thing for timestamp and timestamptz conversions. They should be ISO-8601 (with the `T`). Timestamp must not have any zone offset or additional information, and timestamptz must have a valid zone offset and not a zone name or descriptor.

rdblue replied:
You may want to omit these for now and do them in a follow-up. I think it would make sense to have a set of util methods for date/time like the Java implementation has that are separately tested and used in multiple places.

rdblue commented on python/tests/expression/test_literals.py:
```diff
@@ -0,0 +1,665 @@
+
+
+def test_integer_to_date_conversion():
+    lit = literal(20220303)
```
This is a bit misleading because the integer here looks like a date. But this should actually represent 20 million days from the unix epoch. What this should test is that the integer and literal value are the same object: `lit.value == date_lit.value` and the result is a `DateLiteral`.

rdblue commented on python/tests/expression/test_literals.py:
```diff
@@ -0,0 +1,665 @@
+
+def test_integer_to_date_conversion():
+    lit = literal(20220303)
+    dbl_lit = lit.to(DateType())
```
Did you mean to rename the variable?

rdblue commented on python/tests/expression/test_literals.py:
```diff
@@ -0,0 +1,665 @@
+    lit = literal(34).to(LongType())
+    float_lit = lit.to(FloatType())
+
+    assert math.isclose(lit.value, float_lit.value)
```
Tests should not use `isclose`. All floating point value tests must be exact.

rdblue commented on python/tests/expression/test_literals.py:
```diff
@@ -0,0 +1,665 @@
+
+def test_string_literal():
+    sqrt2 = literal("1.414").to(StringType())
+    phi = literal("3.141").to(StringType())
```
Do you mean `pi`?

rdblue commented on python/tests/expression/test_literals.py:
```diff
@@ -0,0 +1,665 @@
+    assert literal("abc") == literal("abc").to(StringType())
+
+
+def test_string_to_date_literal():
```
These tests should use static values from the spec and Java implementation rather than pulling in fastavro. I know that's what the Java implementation does, but that was to build confidence initially. We don't need to do that here and that removes the need for an extra dependency.

rdblue commented on python/tests/expression/test_literals.py:
```diff
@@ -0,0 +1,665 @@
+    assert avro_val == timestamp.value
+
+
+def test_timestamp_with_zone_without_zone_in_literal():
```
Good!

rdblue commented on python/tests/expression/test_literals.py:
```diff
@@ -0,0 +1,665 @@
+
+
+def test_timestamp_with_zone_without_zone_in_literal():
+    with raises(RuntimeError):
```
Should this be `RuntimeError`? I think a more specific exception would be good so that it can be caught.

rdblue commented:
Thanks, @dramaticlly! This is making great progress.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,601 @@
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value))
+        else:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP))
```
I'm not sure this works. My testing shows that the `quantize` method will increase precision when used this way:

```python
type_var.scale
=> 2

decimal.Decimal(-932).quantize(decimal.Decimal((0, (1,), -type_var.scale)))
=> Decimal('-932.00')

decimal.Decimal(-932).quantize(decimal.Decimal((0, (1,), -type_var.scale))) / 10**type_var.scale
=> Decimal('-9.32')
```

rdblue replied:
Ah, nevermind! I was thinking this was coming from an unscaled value. This works correctly when you use `Decimal("-9.32111")` and set the scale to 2.

dramaticlly commented:
Thank you @samredai  for sending me the patch for addressing most of the comments from @rdblue , I was on personal leave last week so did not have the chance to update in time. 

I believe Sam helped with parametrizing the Literal class with python primitive type, added type hint for both parameters and return type, introduced ABC abstract method, removed several intermediate abstraction together with better handle on float and datetime literal.

`fastavro` is no longer part of the test dependency and rebased from latest upstream to resolve conflict :) 

CC @jun-he

jun-he commented on python/tests/expression/test_literals.py:
```diff
@@ -0,0 +1,662 @@
+def test_invalid_boolean_conversions():
+    assert_invalid_conversions(
+        literal(True),
+        [
```
This list of types seems to be reused in multiple places, can we refactor to define it once?

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+    """
+    if value < FloatType.min or value > FloatType.max:
+        return DoubleLiteral(value)
+    return FloatLiteral(value)
```
This could just return `DoubleLiteral(value)`. If the type needed is a float, then expression binding will convert.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+        >>> literal(123)
+        IntegerLiteral(123)
+    """
+    raise TypeError(f"Unimplemented Type Literal for value: {str(value)}")
```
Should this use `repr` instead of `str` so it is easier to see what was passed in?

dramaticlly replied:
hey sorry, not sure if I followed your thought on this.
this `literal` is the factory method which take the python primitive value and then convert to the iceberg Literal. I feel as the base method of singledispatch, it does not have the access to the concrete instance of parametrized Literal, so I am not sure if we can use `repr` here.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+
+class BooleanLiteral(Literal[bool]):
+    def __init__(self, value):
+        self._value = value
```
Doesn't this need to call `super().__init__(value)` so that it goes through [the check for `None`](https://github.com/apache/iceberg/pull/4262/files#diff-22fa50c420ab8a210f0470d50d942b30397fe4052dab287a8795320bf9b709e1R64-R65)?

rdblue replied:
Or maybe this should just not override `__init__`? Looks like that's what `LongLiteral` does.

samredai replied:
+1 I think removing all of the `__init__` methods and letting it fall back the super class is the way to go.

dramaticlly replied:
Thank you, I removed `__init__` and it shall fallback to super

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+
+class IntegerLiteral(Literal[int]):
+    def __init__(self, value):
+        self._value = value
```
I think all of the init methods need to check for None.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+
+@literal.register(bytearray)
+def _(value: bytearray) -> Literal[bytes]:
+    return BinaryLiteral(value)
```
I'd probably return `BinaryLiteral` for both bytearray and bytes. That's the more common one and expression binding will handle conversion to FixedLiteral if needed.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+    """
+    if value < IntegerType.min or value > IntegerType.max:
+        return LongLiteral(value)
+    return IntegerLiteral(value)
```
Similar to float, I think this should just return `LongLiteral` every time and let expression binding convert to `IntegerLiteral`

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+
+    @to.register(FloatType)
+    def _(self, type_var: FloatType) -> "FloatLiteral":
+        return FloatLiteral(self.value)
```
This is missing the conversion to `float`. Is there a way to add type checking when constructing literals like `FloatLiteral`?

dramaticlly replied:
added `float(value)` here. From what I can tell, the parameter of `FloatLiteral` constructor is set to `float` but one can still pass with int or other type, but its `_value32` with round trip to struct seem to be float

samredai replied:
> Is there a way to add type checking when constructing literals

The [typeguard](https://github.com/agronholm/typeguard) library is a common alternative to explicitly checking the type with if statements, the `@typechecked` decorator in particular is what I've seen used a lot.

rdblue replied:
I just mean that we want type checking to help us catch places where we're passing the wrong type into classes like `FloatLiteral`. I'll look more into this locally.

samredai replied:
Oh I see, I would have expected mypy to catch this when the tests ran since it should be able to deduce that the `T` in `value: T` is an int. It's possible that this might be a gap in mypy where it's not factoring in that LongLiteral inherits from `Literal[int]`.

Typically though mypy will raise an error during tests along the lines of `Argument 1 to "__init__" has incompatible type "int"; expected "float"`.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+        self._value32 = struct.unpack("<f", struct.pack("<f", value))[0]
+
+    def __eq__(self, other):
+        self._value32 == other
```
`_value32` should be used for all comparison methods, not just `__eq__`.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+
+    @to.register(DecimalType)
+    def _(self, type_var: DecimalType) -> DecimalLiteral:
+        dec_val = Decimal(str(self.value))
```
Isn't self.value already a str? I think this may hide bugs because it should be.

dramaticlly replied:
yeah I removed `str` here for now, but I think this might be similar to problem above, I feel for dynamically check language like python its duck-typing nature makes very hard to assume type invariant of value

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+    def _(self, type_var: DecimalType) -> DecimalLiteral:
+        dec_val = Decimal(str(self.value))
+        value_scale = abs(dec_val.as_tuple().exponent)
+        if value_scale == type_var.scale:
```
This block doesn't make sense to me. If the value scale matches, then why does the value need to be modified?

Also, I think this should return the Decimal as-is rather than altering the scale. Here's the Java implementation:

```java
        case DECIMAL:
          // do not change decimal scale
          BigDecimal decimal = new BigDecimal(value().toString());
          return (Literal<T>) new DecimalLiteral(decimal);
```

That is equivalent to `return DecimalLiteral(Decimal(self.value))`

dramaticlly replied:
yeah returning scale from String if scale matches
```
if value_scale == type_var.scale:
    return DecimalLiteral(Decimal(self.value))
```

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+    def _(self, type_var: FixedType) -> "FixedLiteral":
+        if len(self.value) == type_var.length:
+            return self
+        raise TypeError(f"Cannot cast to fixed with different length: {type_var.length}")
```
This should not throw an exception. Instead, this should return None.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+    def _(self, type_var: FixedType) -> FixedLiteral:
+        if type_var.length >= len(self.value):
+            return FixedLiteral(self.value)
+        raise ValueError(f"Cannot cast binary to fixed of smaller size: {type_var.length}")
```
Same here. This should return None if the conversion doesn't succeed.

rdblue commented on python/tests/expression/test_literals.py:
```diff
@@ -0,0 +1,662 @@
+    DateLiteral,
+    FixedLiteral,
+    StringLiteral,
+    TimestampLiteral,
```
`TimeLiteral`?

dramaticlly replied:
it's missing in import but there's unit test for this. I added them with the type checking on top of existing value match

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+
+    @to.register(DateType)
+    def _(self, type_var: DateType) -> "DateLiteral":
+        return DateLiteral((datetime.datetime.strptime(self.value, "%Y-%m-%d") - EPOCH).days)
```
I think it will be more reliable to use `date` rather than `datetime`:

```python
from datetime.datetime import date
EPOCH_DATE = date.fromisoformat("1970-01-01") # OR date(1970, 1, 1)
...
return DateLiteral((date.fromisoformat(self.value) - EPOCH_DATE).days)
```

rdblue replied:
I did some testing for `date.fromisoformat` and it correctly validates the number of digits.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+        )
+
+    @to.register(TimestampType)
+    def _(self, type_var: TimestampType) -> TimestampLiteral:
```
I think there's an easier way to do this:

```python
from datetime import datetime
import re

EPOCH = datetime.fromisoformat("1970-01-01T00:00:00.000000")
ISO_TIMESTAMP = re.compile("\d\d\d\d-\d\d-\d\dT\d\d:\d\d:\d\d(.\d{1,6})?")
...
    if ISO_TIMESTAMP.fullmatch(self.value):
        try:
            delta = (datetime.fromisoformat(self.value) - EPOCH)
            return (delta.days * 86400 + delta.seconds) * 1_000_000 + delta.microseconds
        except TypeError:
            return None
    return None
```

Unfortunately, `fromisoformat` is a bit too permissive, as is `strptime`. The problem with `fromisoformat` is that it accepts partial ISO timestamps, like `"2022-03-28"` and doesn't validate that all fields are there. The problem with `strptime` is that it accepts strings that are missing digits, like `2022-3-28 1:2:3`. That's why I think we should use `fromisoformat`, but strictly validate the incoming timestamp string using a regular expression.

This also uses logic similar to `total_seconds` but doesn't needlessly convert to float.

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+                raise e
+            raise TypeError(f"Cannot convert String to Timestamptz, missing timezone: {self.value}")
+        utc_epoch = EPOCH.replace(tzinfo=pytz.UTC)
+        return TimestampLiteral(int((timestamp - utc_epoch).total_seconds() * 1_000_000))
```
Similar to above, I think we can rewrite this:

```python
from datetime import datetime
import re

EPOCHTZ = datetime.fromisoformat("1970-01-01T00:00:00.000000+00:00")
ISO_TIMESTAMPTZ = re.compile("\d\d\d\d-\d\d-\d\dT\d\d:\d\d:\d\d(.\d{1,6})?[-+]\d\d(:\d\d)?")
...
    if ISO_TIMESTAMPTZ.fullmatch(self.value):
        try:
            delta = (datetime.fromisoformat(self.value) - EPOCHTZ)
            return (delta.days * 86400 + delta.seconds) * 1_000_000 + delta.microseconds
        except TypeError:
            return None
    return None
```

This differs slightly from the `timestamp` version. The regex requires an offset and it uses an epoch value that is in UTC. Also, value returned by `fromisoformat` must both have valid `tzinfo`, which is why this needs `EPOCHTZ`.

rdblue replied:
I've also verified that this works with test cases from Java:
* -1_000_001 <=> "1969-12-31T16:59:58.999999-07:00" and "1969-12-31T23:59:58.999999+00:00"
* 1503091261919314 <-> "2017-08-18T14:21:01.919314-07:00" and "2017-08-18T14:21:01.919314+00:00"

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+            except:
+                raise e
+            raise TypeError(f"Cannot convert String to Timestamptz, missing timezone: {self.value}")
+        utc_epoch = EPOCH.replace(tzinfo=pytz.UTC)
```
I don't think this should modify `EPOCH`. Instead, we should use a different constant.

rdblue commented on python/setup.cfg:
```diff
@@ -42,6 +42,10 @@ package_dir =
 python_requires = >=3.7
+install_requires =
+    pytz
+    python-dateutil
```
I don't think that either one of these libraries is needed. I just did a thorough review of what `datetime` supports and I was able to get everything working.

The main reason why we don't need these is that we are _very_ strict about date/time conversions and only support a limited set of strings in ISO 8601 formats, and support only offsets rather than zone strings.

dramaticlly replied:
removed both `pytz` and `python-dateutil`

rdblue commented on python/src/iceberg/expression/literals.py:
```diff
@@ -0,0 +1,502 @@
+        return DateLiteral((datetime.datetime.strptime(self.value, "%Y-%m-%d") - EPOCH).days)
+
+    @to.register(TimeType)
+    def _(self, type_var: TimeType) -> "TimeLiteral":
```
Similar to the other date/time types, this can be done a bit more easily:

```python
t = time.fromisoformat(self.value)
return (((t.hour * 60 + t.minute) * 60) + t.second) * 1_000_000 + t.microsecond
```

The `time.fromisoformat` method does a good job of validating the format, so no additional regex is needed.

dramaticlly replied:
Thank you Ryan, your example in String to TimeStamp/Date/Time is really helpful

rdblue commented:
Thanks, @dramaticlly!

dramaticlly commented:
Thank you @rdblue  for merging the pull request. Let me also try to add those unresolved comments into issue and will come back with follow up PR.

rdblue commented:
I opened a follow-up PR to fix some of the remaining issues, #4436.

samredai commented:
Woohoo! Big thanks @dramaticlly for tackling this. 🍻
